java/compilers/Symbol.java

```diff
@@ -33,25 +55,14 @@ public String toString() {
         return token;
     }
 
-    static SymbolType matchSymbol(String token) {
-        switch (token) {
-            case "->": return SymbolType.RULE;
-            case "|": return SymbolType.ALT;
-            case "$": return SymbolType.EOF;
-            case "lambda": return SymbolType.LAMBDA;
-            default:
-                if (token.matches("[a-z]+"))
-                    return SymbolType.TERMINAL;
-                if (token.matches("[A-Z]+"))
-                    return SymbolType.NONTERMINAL;
-                return null;
-        }
-    }
-
     @Override
     public boolean equals(Object o) {
-        if (this == o) return true;
```

Comment:
The reason behind this is that `Symbol.EOF.equals(null)` will be true, which is desirable since the input queue will return null when all input is consumed. It would technically break the contract between `equals` and `hashCode` since null has no hash code, but helps simplify the parsing.

Reply:
So this is a solved problem we've discussed in class.  IIRC Hellman mentioned that the input queue is always padded with some number of EOF tokens.  That's the way to solve this problem, not a hack like this.  I'm also not positive that this code gets run the way you're imagining.
There are a couple of places where you are equality comparing instances of the symbol class rather than the `SymbolType` enum (line 61 here is one you added and in ParseTree.java:36 you compare a symbol with Symbol.LAMBDA which is shady for the same reasons). This seems like bad design to me, and I think it will eventually cause a bug.
This is why I prefer using .equals on enums because you're still guaranteed no NPE and it's harder to mix up the semantics between the two types), and totally sidestepping the `Symbol#equals` function.
Maybe in this case you know from the static factory method and the code you've written which uses this class that there will never be any dissonance between reference and identity, maybe you sidestepped the equals logic intentionally in those places, but the members on this class aren't final, and there's nothing to stop the reference and value semantics from getting out of sync.  There's also now an empty constructor that didn't exist before, for a class which I would picture as immutable.
When I picture the Symbol class it has a couple of key requirements:
1) it's immutable.  the class is constructed based on an object that exists in the parsed source code or in our grammar, and because of that can't be changed.
2) it's comparable using equals.  we don't want to have to track the places a special value is created (like EOF) to make sure that our identity comparisons are always to the right instances ones.  (imagine someone just makes an empty Symbol, and populates it with token="$", type=SymbolType.EOF, then ParseTree.java:40 might fail sometimes)
3) it's not nullable.  if it's absolutely necessary there could be a special value for 'empty symbol' or something like that, but if it's possible this should be avoided imo (and so should null).  Ideally I think `Symbol#of` would return an `Optional<Symbol>`, and that way an error constructing a symbol is an explicit absence of a value, not null (which is ambiguous).

tl;dr there better ways to solve this problem, there are ways to get better language-level guarantees of code correctness by constraining this class more, the current version seems error-prone to me based on my java experience on very large projects.

Reply:
Yes this is valid, it was meant as a simple fix and something I can work on later. Quoted in #6 for future reference.